File: app/src/main/java/com/leodeleon/popmovies/feature/viewModel/MovieDetailsViewModel.java
package com.leodeleon.popmovies.feature.viewModel;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import com.leodeleon.popmovies.data.MovieRepository;
import com.leodeleon.popmovies.model.Movie;
import com.leodeleon.popmovies.model.MovieDetail;
import io.reactivex.CompletableObserver;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.subjects.PublishSubject;
import java.util.List;
import javax.inject.Inject;
import javax.inject.Singleton;
import timber.log.Timber;

@Singleton
public class MovieDetailsViewModel extends ViewModel {

  private final MutableLiveData<MovieDetail> detailLiveData = new MutableLiveData<>();
  private final MutableLiveData<List<String>> videoLiveData = new MutableLiveData<>();
  private final MovieRepository movieRepository;
  private final CompositeDisposable disposable = new CompositeDisposable();
  private final PublishSubject<Movie> movieSubject;

  @Inject
  public MovieDetailsViewModel(MovieRepository repository, PublishSubject<Movie> subject) {
    movieRepository = repository;
    movieSubject = subject;
  }

  @Override protected void onCleared() {
    super.onCleared();
    disposable.clear();
  }

  public void loadDetails(int id) {
    Disposable d1 = movieRepository.getMovieDetail(id).subscribe(detailLiveData::postValue);
    Disposable d2 = movieRepository.getVideoKeys(id).subscribe(videoLiveData::postValue);
    disposable.add(d1);
    disposable.add(d2);
  }

  public MutableLiveData<MovieDetail> getDetailLiveData() {
    return detailLiveData;
  }

  public MutableLiveData<List<String>> getVideoLiveData() {
    return videoLiveData;
  }

  public void addFavorite(Movie movie) {
    movie.setIsFavorite(1);
    movieRepository.saveMovie(movie).subscribe(new MovieCompletableObserver(movie));
  }

  public void removeFavorite(Movie movie) {
    movie.setIsFavorite(0);
    movieRepository.saveMovie(movie).subscribe(new MovieCompletableObserver(movie));
  }

  private class MovieCompletableObserver implements CompletableObserver {
    private Movie movie;

    MovieCompletableObserver(Movie movie) {
      this.movie = movie;
    }

    @Override public void onSubscribe(@NonNull Disposable d) {
      disposable.add(d);
    }

    @Override public void onComplete() {
      movieSubject.onNext(movie);
    }

    @Override public void onError(@NonNull Throwable e) {
      Timber.e(e);
    }
  }
}

File: app/src/main/java/com/leodeleon/popmovies/feature/viewModel/MovieDetailsViewModel.kt
package com.leodeleon.popmovies.feature.viewModel

import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.leodeleon.popmovies.data.MovieRepository
import com.leodeleon.popmovies.model.Movie
import com.leodeleon.popmovies.model.MovieDetail
import io.reactivex.CompletableObserver
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Consumer
import io.reactivex.subjects.PublishSubject
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class MovieDetailsViewModel @Inject
constructor(
    private val movieRepository: MovieRepository,
    private val movieSubject: PublishSubject<Movie>
) : ViewModel() {

  val detailLiveData = MutableLiveData<MovieDetail>()
  val videoLiveData = MutableLiveData<List<String>>()
  private val disposable = CompositeDisposable()

  override fun onCleared() {
    super.onCleared()
    disposable.clear()
  }

  fun loadDetails(id: Int) {
    val d1 = movieRepository.getMovieDetail(id).subscribe(Consumer<MovieDetail> { detailLiveData.postValue(it) })
    val d2 = movieRepository.getVideoKeys(id).subscribe(Consumer<List<String>> { videoLiveData.postValue(it) })
    disposable.add(d1)
    disposable.add(d2)
  }

  fun addFavorite(movie: Movie) {
    movie.favorite = 1
    movieRepository.saveMovie(movie).subscribe(MovieCompletableObserver(movie))
  }

  fun removeFavorite(movie: Movie) {
    movie.favorite = 0
    movieRepository.saveMovie(movie).subscribe(MovieCompletableObserver(movie))
  }

  private inner class MovieCompletableObserver
  internal constructor(private val movie: Movie) : CompletableObserver {

    override fun onSubscribe(d: Disposable) {
      disposable.add(d)
    }

    override fun onComplete() {
      movieSubject.onNext(movie)
    }

    override fun onError(e: Throwable) {
      Timber.e(e)
    }
  }
}
